madara/crates/primitives/bloom_filter/src/events.rs

```diff
@@ -0,0 +1,487 @@
+    ///
+    /// # Returns
+    ///
+    /// A new EventBloomWriter initialized with the provided events
+    pub fn from_events<'a, I>(events: I) -> Self
+    where
+        I: Iterator<Item = &'a Event> + 'a,
+    {
+        let mut unique_keys = HashSet::new();
+
+        for key in Self::events_to_bloom_keys(events) {
+            unique_keys.insert(key);
+        }
+
+        let size = (unique_keys.len() as f64 * BITS_PER_ELEMENT).ceil() as _;
+
+        let filter = BloomFilter::<_, AtomicBitStore>::new(size, HASH_COUNT);
+
+        for key in unique_keys {
+            filter.add(&key);
+        }
+        Self { filter }
+    }
+
+    pub fn size(&self) -> u64 {
+        self.filter.size()
+    }
+
+    fn events_to_bloom_keys<'a, I>(events: I) -> impl Iterator<Item = [u8; 33]> + 'a
+    where
+        I: Iterator<Item = &'a Event> + 'a,
+    {
+        events.flat_map(Self::event_to_bloom_keys)
+    }
+
+    /// Converts a single event into an iterator of bloom keys.
+    ///
+    /// # Arguments
+    ///
+    /// * `event` - A reference to an Event struct
+    ///
+    /// # Returns
+    ///
+    /// An iterator yielding 33-byte arrays, one for from_address and one for each event key
+    fn event_to_bloom_keys(event: &Event) -> impl Iterator<Item = [u8; 33]> + '_ {
+        let from_address_key = create_bloom_key(0, &event.from_address);
+
+        let keys = event.keys.iter().enumerate().map(|(i, key)| create_bloom_key(i as u8 + 1, key));
+
+        std::iter::once(from_address_key).chain(keys)
+    }
+}
+
+/// Reader component for querying event-based Bloom filters.
+///
+/// The reader provides an immutable view into a Bloom filter, typically
+/// deserialized from a previously constructed [`EventBloomWriter`].
+///
+/// # Serialization
+///
+/// The reader implements [`Deserialize`] to allow efficient reconstruction
+/// from serialized data.
+///
+/// # Memory Usage
+///
+/// The reader's memory footprint is minimal, consisting only of:
+/// - The bit array (size determined by the writer)
+/// - A small fixed overhead for the filter metadata
+pub struct EventBloomReader {
+    filter: BloomFilter<BloomHasher, BitStore>,
+}
+
+impl<'de> Deserialize<'de> for EventBloomReader {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let storage = BitStore::deserialize(deserializer)?;
+
+        Ok(Self { filter: BloomFilter::from_storage(storage, HASH_COUNT) })
+    }
+}
+
+/// Pattern matching component for searching event-based Bloom filters.
+///
+/// The searcher supports complex queries with both AND and OR semantics:
+/// - Different patterns are combined with AND logic
+/// - Keys within each pattern use OR logic
+///
+/// # Query Optimization
+///
+/// The searcher pre-calculates hashes for all patterns during construction,
+/// making subsequent searches more efficient.
+pub struct EventBloomSearcher {
+    patterns: Vec<Vec<PreCalculatedHashes>>,
```

Comment:
I think in theory you can OR the bits from the bloom keys to emulate the AND semantic right?
if your first key is 0100000 and your second key is 0001000, you can bitwise or the two filters to get a filter that tests for the first and second key at once, or am i missing something?